### PR TITLE
Add Trivy scanning to the build pipeline

### PR DESCRIPTION
**Pull request by SoapHia**
Adding Trivy scanning to Cloud Build involved adding a few extra steps to cloudbuild.yaml. The resulting Cloud Build will:

1. Scan the new image after it has been built from a release
2. Store results in table and json format into CI workspace
3. Parse results from the json output from (2) using a Python script, and create a GitHub issue containing the table output from (2) if any vulnerabilities were found (See SoapHia#7 for an example)
4. If vulnerabilities with CRITICAL/HIGH severity were found, the build will fail and prevent the images from being pushed to GCR.

For simplicity, at most one issue per release will be created, containing all the vulnerabilities found in that release.

### CODE REVIEW

File: ci/parse-trivy-results.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+"""Script to parse results from Trivy scan.
```
Comment (Temikus) [inline ci/parse-trivy-results.py:1]: You seem to have a whole bunch of trailing spaces in this file, please fix:
```
ci/parse-trivy-results.py:7:70: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:9:68: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:103:64: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:105:0: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:108:74: C0303: Trailing whitespace (trailing-whitespace)
```

I suggest setting your editor to automatically trim whitespace.

Comment (Temikus) [inline ci/parse-trivy-results.py:1]: You seem to have a whole bunch of trailing spaces in this file, please fix:

ci/parse-trivy-results.py:7:70: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:9:68: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:103:64: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:105:0: C0303: Trailing whitespace (trailing-whitespace)
ci/parse-trivy-results.py:108:74: C0303: Trailing whitespace (trailing-whitespace)
I suggest setting your editor to automatically trim whitespace.

File: ci/parse-trivy-results.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+Usage:
+    python3 parse-trivy-results.py <TRIVY_FILE> <COMMIT_ID> <BUILD_ID> <FULL_TAG> <REPO_NAME> <TOKEN_FILE>
+
+"""
+import json
+import sys
+from github import Github
+
+CMDLINE_ARGS = 7
```
Comment (Temikus) [inline ci/parse-trivy-results.py:22]: Nit: Why is this outlined as CAPS? Ditto for the ones below.

File: ci/parse-trivy-results.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+Both the json and table Trivy results must be created prior to running 
+this script. These can be generated with:
+    trivy image --format json --light --no-progress -o <OUTPUT_NAME> 
+        gcr.io/cre-tools/ts-bridge
+    trivy image --light  --no-progress -o <OUTPUT_NAME>
+        gcr.io/cre-tools/ts-bridge
+
+Usage:
+    python3 parse-trivy-results.py <TRIVY_FILE> <COMMIT_ID> <BUILD_ID> <FULL_TAG> <REPO_NAME> <TOKEN_FILE>
```
Comment (Temikus) [inline ci/parse-trivy-results.py:15]: Add example to clarify format of TRIVY_FILE, COMMIT_ID, REPO_NAME, FULL_TAG and TOKEN_FILE

File: ci/parse-trivy-results.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    """Connects to GitHub API and returns a repo object."""
+    REPO_NAME = sys.argv[5]
+    with open(sys.argv[6]) as f:
+        token = f.read()
+    github = Github(token)
+    return github.get_repo(REPO_NAME)
+
+def create_issue(target_name, num_vulnerabilities, severity_list, table):
+    """Creates a Github issue with vulnerabilities as description."""
+    COMMIT_ID = sys.argv[2]
```
Comment (Temikus) [inline ci/parse-trivy-results.py:60]: This seems relatively unsafe parsing-wise as it's easy to mistake the order of a giant list of options and then there's no validation on the input.

- Please use `flags`.
- Please validate input. E.g. check if file exists, check if formatting is correct, etc.

Feel free to ping me to discuss.

File: ci/parse-trivy-results.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        num_vulnerabilities = len(vulnerabilities)
+        severity_list = get_severity_list(vulnerabilities)
+        issue_number = create_issue(target, num_vulnerabilities, 
+                                    severity_list, trivy_table)
+        
+        debug_msg = ("{} vulnerabilities of type: [{}] were found in image. "
+                     "Please refer to issue: {} for details."
+                    ).format(num_vulnerabilities, ",".join(severity_list), 
+                             issue_number)
+        print(debug_msg)
```
Comment (Temikus) [inline ci/parse-trivy-results.py:110]: I think you may have left a debug statement here.

**Review by Temikus — CHANGES_REQUESTED** [review]

File: ci/parse-trivy-results.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+import sys
+from github import Github
+
+CMDLINE_ARGS = 7
+
+def load_results():
+    """ Load the results from Trivy."""
+    trivy_out_json = "{}.json".format(sys.argv[1])
+    trivy_out_table = "{}.table".format(sys.argv[1])
+    print(trivy_out_table, trivy_out_json)
```
Comment (Temikus) [inline ci/parse-trivy-results.py:31]: Give a bit more details on what's printing here so it's easier to understand what's going on in the build log

File: ci/parse-trivy-results.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    repo = get_github_repo()
+
+    title = ("Vulnerability [{}] found in {}: Images from commit {} cannot be "
+             "released").format(",".join(severity_list), release_tag, commit_id)
+
+    intro = ("Trivy has detected {} vulnerabilities in your latest "
+             "build. Please correct this issue so the new images can be "
+             "published on Container Registry.\n").format(num_vulnerabilities)
+    body = [intro]
+    body.append("**Cloud Build ID:** {}".format(build_id))
```
Comment (Temikus) [inline ci/parse-trivy-results.py:75]: Does the long brace format work here? E.g.:

```
body.append ("**Commit ID:** {}".format(commit_id)
                        "**Tag:** {}".format(release_tag)
                        "**Target:** {}".format(target_name))
```         

Not entirely sure how pythonic this is so you can ask the readability reviewer too.

File: ci/parse-trivy-results.py (hunk 1 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py:1]: Fixed :)

File: ci/parse-trivy-results.py (hunk 2 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: This was accidentally left from a previous version of the file, I've fixed it now :)

File: ci/parse-trivy-results.py (hunk 4 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: As discussed offline, I've added absl to check the cmdline args.

File: ci/parse-trivy-results.py (hunk 5 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: As discussed offline, "debug" is a misleading name here. Its purpose is to provide extra details in the build log. I have change the variable name to "details" to address this.

File: ci/parse-trivy-results.py (hunk 6 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: This was a debug message which I forgot to remove. I have added comments in the lines following to provide more context to trivy results

File: ci/parse-trivy-results.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        gcr.io/cre-tools/ts-bridge
+    trivy image --light  --no-progress -o <trivy_output>.table
+        gcr.io/cre-tools/ts-bridge
+"""
+import json
+import sys
+from absl import app
+from absl import flags
+from github import Github
+from pathlib import Path
```
Comment (jaqx0r) [inline ci/parse-trivy-results.py:19]: I'm used to seeing "import github" and "import pathlib" and then using the form 'github.Github' and 'pathlib.Path' when you call those methods.  That's Google internal style though I'm not sure what public style requires.

File: ci/cloudbuild.yaml [hunk 9]
```diff
@@ -31,14 +31,53 @@ steps:
...
+    pip install PyGithub
+    pip install absl-py
+    python3 ci/parse-trivy-results.py \
+      --build_id=-$BUILD_ID \
+      --commit_id=$SHORT_SHA \
+      --release_tag=$$(cat _release_tag) \
+      --repo_name=$_REPO_NAME \
+      --token_file=git_token.txt \
+      --trivy_file=trivy-out
+# TODO: change repo name to Google/
```
Comment (jaqx0r) [inline ci/cloudbuild.yaml:78]: Does this need to happen in this PR?

**Review by jaqx0r — CHANGES_REQUESTED** [review]

File: ci/parse-trivy-results.py (hunk 3 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: This has been replaced by absl.flags

File: ci/parse-trivy-results.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    severity_list = [v.get("Severity") for v in vulnerabilities]
+    return sorted(set(severity_list))
+
+
+def get_github_repo():
+    """Connects to GitHub API and returns a repo object."""
+    with open(FLAGS.token_file) as f:
+        token = f.read()
+    github = Github(token)
+    return github.get_repo(FLAGS.repo_name)
```
Comment (Temikus) [inline ci/parse-trivy-results.py:102]: We should fail gracefully if `get_repo` fails.

File: ci/parse-trivy-results.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    body.append("**Cloud Build ID:** {}".format(FLAGS.build_id))
+    body.append("**Commit ID:** {}".format(FLAGS.commit_id))
+    body.append("**Tag:** {}".format(FLAGS.release_tag))
+    body.append("**Target:** {}".format(target_name))
+    body.append("```")
+    body.append(table)
+    body.append("```")
+    body = "\n".join(body)
+
+    new_issue = repo.create_issue(title=title, body=body)
```
Comment (Temikus) [inline ci/parse-trivy-results.py:132]: What happens if creating an issue fails? Ideally we should fail gracefully on it.

**Review by Temikus — CHANGES_REQUESTED** [review]

File: ci/parse-trivy-results.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+                  "build.").format(num_vulnerabilities)
+
+    if high_or_critical_exists(severity_list):
+        title += ": Images from commit {} cannot be released".format(
+            FLAGS.commit_id)
+
+        body_intro += " Please correct this issue so the new images can be "
+        "published on Container Registry."
+
+    body = [body_intro]
```
Comment (jaqx0r) [inline ci/parse-trivy-results.py:122]: As discussed elsewhere, body = """Please ...
Cloud Build ID: {build_id}
Commit: {commit}
""".format(dict(build_id=FLAGS.build_id, etc etc ))

File: ci/parse-trivy-results.py (hunk 8 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py:19]: As discussed offline, this is the style used in https://github.com/PyGithub/PyGithub, so we will keep it unchanged.

File: ci/parse-trivy-results.py (hunk 10 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: Added exception handling to address this

File: ci/parse-trivy-results.py (hunk 11 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: Added exception handling to address this.

File: ci/parse-trivy-results.py (hunk 12 above)
Comment (SoapHia) [reply ci/parse-trivy-results.py]: done, please resolve this comment if the new style looks ok :)

**Review by jaqx0r — APPROVED** [review]

File: ci/cloudbuild.yaml (hunk 9 above)
Comment (SoapHia) [reply ci/cloudbuild.yaml]: It was kept as the forked branch for testing purposes, I have updated it to be the google branch instead.

**Pull request merged**